### ISSUE TITLE
PID file no longer gets created

### ISSUE THREAD

**Issue by windware** [vstakhov/rspamd#2166]
### Classification (Please choose *one* option):

* [ ] Crash/Hang/Data loss
* [ ] WebUI/Usability
* [ ] Serious bug
* [X] Other bug
* [ ] Feature
* [ ] Enhancement

### Reproducibility (Please choose *one* option):

* [X] Always
* [ ] Sometimes
* [ ] Rarely
* [ ] Unable
* [ ] I didn’t try
* [ ] Not applicable

### Rspamd version:
1.7.3 (1.7.3-1~xenial)

### Operation system, CPU, memory and environment:
Ubuntu 16.04 64bit

### Description (Please provide a descriptive summary of the issue):
It no longer creates the PID file with the log saying,
> main: skip writing pid in no-fork mode

even when specified in 'rspamd.conf' under 'options' -> 'pidfile'.

### Steps to reproduce:
Start rspamd.

### Expected results:
PID file should still be created when not forking.

### Actual results:
PID file no longer gets created.

### Additional information:

It became a problem as I could no longer monitor rspamd by checking on its PID file.

To get the PID file, systemd launch script has to be changed by taking '-f' option out of 'ExecStart' and add 'Type=forking'.

**Comment by windware** [comment]
I still would like to have PID file support as every time I update the rspamd package, it reverts the changes I made in systemd config file and the monitoring complains that the service cannot be found as no longer a PID file is written.

I do have options such as monitoring by process name or with the above 'export' command, write down the PID to a file in an interval but having an official PID file is the cleanest solution.

**Comment by caffeinejolt** [comment]
Agreed - having a PID file for systems like monit would be very useful.

**Comment by julian70400** [comment]
Same problem here on Debian 10 buster, no way to monitor Rspamd with Monit..

**Comment by vstakhov** [comment]
Is it anyhow related to Rspamd? Are you using the packages that I provide?

**Comment by julian70400** [comment]
I currently use : http://rspamd.com/apt-stable/ buster main
rspamd works like a charm except it do not generate pid file (which, as I said, making troubles for monitoring).
If you have any ideas, I am interested.

**Comment by DamonFstr** [comment]
+1 on having a PID file for systems like monit

**Comment by julian70400** [comment]
Another way to work with monit without PID is a config like this : 

```
check process rspamd
    matching 'rspamd: main process'
    start program = "/sbin/service rspamd start"
    stop program = "/sbin/service rspamd stop"

   if cpu is greater than 40% for 2 cycles then alert
   if cpu > 60% for 5 cycles then alert
   if memory > 80% for 4 cycles then alert
   if totalmem > 512 MB for 5 cycles then alert
```

Works like a charm !

**Comment by ws-hono** [comment]
What is the status of this issue?
While the above comment's method works, finding a process by its name isn't the cleanest solution and I still would like to have a PID file generated.

**Comment by fevarus-design** [comment]
hey guys.. i faced the same issue. so i wrote some bash script which adds pid´s. you can find a tutorial here:

https://www.fevarus-design.de/tipps-und-tricks/bash-scripts-pids-fuer-monit-erstellen/

just create a script.sh in monit or elsewhere and execute it via cron or via monit. if you choose monit, you can set an alert or anything else you like.

you also can extend it for any service you like. just add a new function. i made 2 as an example. otherwise just use the "check process" ... "matching" ... for avoiding a pid.

hope this helps.

**Comment by thorro** [comment]
I have the same problem, logrotate script still wants to read the pidfile after rotation and it fails. Logfiles are inconsistent. To force the logfile reopen, I have modified the logrotate script as follows:

```
...
    postrotate
        kill -USR1 $(systemctl -p MainPID show rspamd.service | sed 's/.*=//')
```

**Comment by fatalbanana** [comment]
Related to #2118

It's confusing to have `pidfile` setting ignored. It would be useful to at least have a toggle available to force it to be still written in foreground mode.

**Comment by vstakhov** [comment]
I'm still not sure about what is the problem. Could you please elaborate more on that?

**Comment by fatalbanana** [comment]
If Rspamd is run in foreground mode (as usual under systemd) then no pidfile is written, though it's defined in config & otherwise used.

The pidfile is not strictly needed but may still be useful to have.
